### DISCUSSION TITLE
Memory management Wasm/Node.js

### DISCUSSION THREAD

**Discussion by tndrle** [emscripten-core/emscripten#18330] [Q&A]
I have a question about memory management. I'm targeting Node.js but I don't know if that is relevant.

Assume I have two C functions which I compile to Wasm with emscripten:
* `open()`: basically allocates memory 
* `close(handle)`: frees the allocated memory

I call `open()` in JavaScript when I create a new instance of the class `Resource`.
```js
const open = Module.cwrap("open", "number", []);
const close = Module.cwrap("close", "number", ["number"]);

class Resource {
  constructor() {
    this.handle = open();
  }

  finalize() {
    close(this.handle);
  }
}

const res = new Resource();
```

When I don't need `res` anymore, I have to call `res.finalize()` to free the memory.
If I don't call `res.finalize()`, I create a memory leak.

Recently, [`FinalizationRegistry`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry) was introduced. Instead of manually calling `res.finalize()`, I could do something like
```js
const registry = new FinalizationRegistry((handle) => {
  close(handle);
});

registry.register(res, res.handle);
```

When the garbage collector (GC) reclaims `res`, it also leads to a call to `close()` (possibly much later, though).
This would free the allocated memory at some point. I don't really care for when exactly this happens, so it should be fine.

However, what I'm wondering is whether the memory allocated in `open()` impacts the GC at all. Or can it happen that the GC is not triggered, because there are only very few unreachable JavaScript objects (GC thinks it's not worthwhile running), even though `open()` has allocated vast amounts of memory? After all, the GC does not know the connection between the allocated memory and `res`, i.e. it doesn't know that it could free memory by just reclaiming `res`.

I hope it is more or less clear what I'm struggling with.
Any thoughts are appreciated.

**Comment by sbc100** [comment]
No, the GC doesn't know about any allocations in the wasm linear memory, and doesn't know that close() free's memory or that open() allocates it.   In other words, its can't/doesn't reason about the links between GC objects and regions of linear memory.

That doesn't mean it won't run though.  Trying to second guess if/when GC might run is probably not worth its since its no something is observable or to be relied upon.

**Reply by tndrle** [reply]
Thanks for the insights
